src/ripple/basics/Slice.h

```diff
@@ -34,34 +37,55 @@
 
 namespace ripple {
 
-/** An immutable linear range of bytes.
+/** A linear range of bytes.
 
     A fully constructed Slice is guaranteed to be in a valid state.
     A Slice is lightweight and copyable, it retains no ownership
     of the underlying memory.
 */
-class Slice
+template<bool Mutable>
```

Comment:
Another place I'd prefer an enum... `Slice<true> Slice<false>` vs. `Slice<Immutable> Slice<Mutable>` You have typedefs so it's less of an issue for code readability, but in the debugger you'll see `Slice<true> Slice<false>` and have to remind yourself what those mean. I *think* (at least in MSVC) you'll get readable enum values in the debugger.

Reply:
Addressed in 91c6c712b0973af30980b849d50e33a4f715d449

I made the change, but I'm ambivalent about it. We need to namespace the enum somehow - either with a `enum class` or with a namespace. So it's going to be `SliceImpl<true>` vs. `SliceImpl<SliceImplMutability::yes>`. If you use an `enum class`, the parameter will not be convertable to a bool. This clutters up the `enable_if` code. Before:
```
        class = std::enable_if_t<
            !Mutable &&
            std::is_same<T, SliceImpl<true>>::value>>
```
After
```
        class = std::enable_if_t<
            Mutable == SliceImplMutability::no &&
            std::is_same<T, SliceImpl<SliceImplMutability::yes>>::value>>
```
We can get around the conversion to bool by defining a `constexpr` class member, so that helps (and I did so). The `no/yes` enum is fine for this (this class is likely to go away), though I'm not convinced we want to do that all the time for boolean template parameters.

BTW, I didn't do `mutable`, `immutable` as the types because `mutable` is a keyword and our enums are lowercase.

Reply:
You can always mark my comments "won't fix" if you don't think it's useful/too much work for too little gain :) bool input params for templates + functions are just a pet-peeve of mine. I've just seen too many bugs from it and wasted too much time debugging/reading code with them. 

I'm almost thinking there could be a pattern of giving names to these kind of bools and keeping them generic (not specific to one class). `enum class IsMutable{ no, yes }; enum class IsAdmin { no, yes }; enum class IsHumanReadable{ no, yes };` etc. Not a perfect solution, but maybe slightly better?